controller/backing_image_manager_controller.go

```diff
@@ -588,9 +588,38 @@ func (c *BackingImageManagerController) deleteInvalidBackingImages(bim *longhorn
 			}
 			return err
 		}
-		if bi.Status.UUID == bim.Spec.BackingImages[biName] {
+
+		// Delete the file from a backing image manager when:
+		//   1. The spec record is removed or does not match.
+		//   2. The file state is failed and there are available files in
+		//      other backing image managers.
+		deleteRequired := false
+		if bi.Status.UUID != bim.Spec.BackingImages[biName] {
+			deleteRequired = true
+		}
+		if !deleteRequired {
+			// Prefer to use the file info in BackingImage.Status
+			fileState := bim.Status.BackingImageFileMap[biName].State
+			if bi.Status.DiskFileStatusMap[bim.Spec.DiskUUID] != nil {
+				fileState = bi.Status.DiskFileStatusMap[bim.Spec.DiskUUID].State
+			}
+			if fileState == longhorn.BackingImageStateFailed {
+				for diskUUID, biFileInfo := range bi.Status.DiskFileStatusMap {
+					if diskUUID == bim.Spec.DiskUUID {
+						continue
+					}
```

Comment:
Why do we skip if `diskUUID == bim.Spec.DiskUUID`?

Reply:
There is no need to re-check the current file, since we already know it's state failed. But this one is unnecessary, I can remove it to avoid confusion.